### PR TITLE
Update AutoExtern from the `compiler-bootstrap` branch

### PR DESCRIPTION
**Pull request by atomb**
This PR pulls in all of the changes to `AutoExtern` from the `compiler-bootstrap` branch, which will leave that branch with only two commits that differs from `master`.

https://github.com/dafny-lang/dafny/commit/f4d915b2a694d49110eb71342d037e656450edde
https://github.com/dafny-lang/dafny/commit/8206206e31f830d765d32f7be2f2fed3476ee03f

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/AutoExtern.Test/Tests.cs [hunk 1]
```diff
@@ -24,9 +24,12 @@ public class IntegrationTests {
     startInfo.ArgumentList.Add("NS"); // Namespace
     startInfo.ArgumentList.Add("Library.dfy.template"); // Template
     startInfo.ArgumentList.Add("CSharpModel.dfy"); // Target for CSharpModel.dfy
+    startInfo.ArgumentList.Add("--skip-interface");
+    startInfo.ArgumentList.Add("NS.ICanClone");
     startInfo.ArgumentList.Add("Library.dfy"); // Target file
     startInfo.ArgumentList.Add("Library1.cs"); // Source file
     startInfo.ArgumentList.Add("Library2.cs"); // Source file
+    startInfo.ArgumentList.Add("Library3.cs"); // Source file
```
Comment (MikaelMayer) [inline Source/AutoExtern.Test/Tests.cs:32]: Where is this Libary3.cs coming from?

Comment (atomb) [reply Source/AutoExtern.Test/Tests.cs:32]: Oh, interesting. I didn't write this code myself, so I'm not sure, but I notice an earlier commit that renamed `Library1.cs`, at least. This test may need to be updated. I'll investigate.

**Review by jtristan — COMMENTED** [review]
Why are we making modifications to the parser? Are these modifications made specifically for DinD Repl's semantics?

**Comment by MikaelMayer** [comment]
> Why are we making modifications to the parser? Are these modifications made specifically for DinD Repl's semantics?

Yes, to accommodate parsing of single expressions and include files.

**Comment by atomb** [comment]
> > Why are we making modifications to the parser? Are these modifications made specifically for DinD Repl's semantics?
> 
> Yes, to accommodate parsing of single expressions and include files.

Those changes are isolated to a single commit that could be left out of this PR.

I'd argue that perhaps it's useful to have them in the main code base (because there are other use cases for parsing single expressions, too), but I'm not strongly attached to that idea.

**Comment by jtristan** [comment]
Are we certain that this modification to the parser cannot impact the definition of a valid program outside of the REPL? For example, would a dafny program with just a single expression in it now be accepted by Dafny?

**Comment by atomb** [comment]
> Are we certain that this modification to the parser cannot impact the definition of a valid program outside of the REPL? For example, would a dafny program with just a single expression in it now be accepted by Dafny?

I can confirm that a single expression in a file is not accepted by Dafny after this change.

File: Source/DafnyCore/Dafny.atg [hunk 2]
```diff
@@ -161,26 +165,41 @@ public static int Parse (string/*!*/ s, string/*!*/ fullFilename, string/*!*/ fi
 }
 
 public static Parser SetupParser(string/*!*/ s, string/*!*/ fullFilename, string/*!*/ filename, Include include, ModuleDecl module,
-                                 BuiltIns builtIns, Errors/*!*/ errors, bool verifyThisFile=true, bool compileThisFile=true) {
+                                 BuiltIns builtIns, Errors/*!*/ errors, bool verifyThisFile=true, bool compileThisFile=true,
```
Comment (MikaelMayer) [inline Source/DafnyCore/Dafny.atg:168]: Is there a test that can cover these changes to the grammar? If not, can you please add one?

Comment (MikaelMayer) [reply Source/DafnyCore/Dafny.atg:168]: Either way are fine for me. I know adding a new test is cumbersome so I understand if you want to remove this extra code. I'm happy to review the PR from AutoExterns that will also contain the tests.

**Comment by atomb** [comment]
I reverted the parser commit, as it really belongs in a separate PR, and could use a test, as @MikaelMayer pointed out.

**Review by jtristan — APPROVED** [review]

**Pull request merged**